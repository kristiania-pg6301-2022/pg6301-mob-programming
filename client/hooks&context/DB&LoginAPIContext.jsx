import React from "react";

export const DBLoginAPIContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
});

export async function fetchJSON(url) {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to load ${res.status}: ${res.statusText}`);
  }
  return await res.json();
}
export async function postJSON(url, object) {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify(object),
  });
  if (!res.ok) {
    throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
  }
}
