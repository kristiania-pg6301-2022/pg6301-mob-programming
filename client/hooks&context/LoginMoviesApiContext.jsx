import React from "react";

export const LoginMoviesApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async listMovies(query) {
    return await fetchJSON("/api/movies?" + new URLSearchParams(query));
  },
  async createMovie(movie) {
    return await postJSON("/api/movies", movie);
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
  userinfo: undefined,
});

export async function fetchJSON(url) {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to load ${res.status}: ${res.statusText}`);
  }
  return await res.json();
}
export async function postJSON(url, object) {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify(object),
  });
  if (!res.ok) {
    throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
  }
}
