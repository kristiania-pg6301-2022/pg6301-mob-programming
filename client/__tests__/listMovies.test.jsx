import { act, Simulate } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { LoginMoviesApiContext } from "../hooks&context/LoginMoviesApiContext";
import MovieLists from "../pages/movieLists";
import React from "react";

const movies = [
  {
    title: "My movie",
    plot: "Something happens",
    poster: "https://example.com/poster1.jpg",
  },
  {
    title: "My other movie",
    plot: "Nothing happens",
    poster: "https://example.com/void1.jpg",
  },
];

async function renderListMovies(listMovies) {
  const element = document.createElement("div");
  await act(async () =>
    ReactDOM.render(
      <LoginMoviesApiContext.Provider value={{ listMovies }}>
        <MovieLists />
      </LoginMoviesApiContext.Provider>,
      element
    )
  );
  return element;
}

describe("list movies", () => {
  it("should show loading screen", async () => {
    const element = await renderListMovies(() => new Promise(() => {}));
    expect(element.querySelector(".loading-indicator")).not.toBeNull();
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show queries by country", async () => {
    const listMovies = jest.fn(() => []);
    const element = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <LoginMoviesApiContext.Provider value={{ listMovies }}>
          <MovieLists />
        </LoginMoviesApiContext.Provider>,
        element
      );
    });
    Simulate.change(element.querySelector("#country-query"), {
      target: { value: "Norway" },
    });
    await act(async () => {
      await Simulate.submit(element.querySelector("form"));
    });
    expect(listMovies).toHaveBeenCalledWith({
      country: "Norway",
    });
  });

  it("should movie list", async () => {
    const element = await renderListMovies(async () => movies);
    expect(element.querySelector("h3").innerHTML).toEqual(movies[0].title);
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show error message", async () => {
    const element = await renderListMovies(async () => {
      throw new Error("Failed to fetch");
    });
    expect(element.querySelector(".error-message").innerHTML).toContain(
      "Failed to fetch"
    );
    expect(element.innerHTML).toMatchSnapshot();
  });
});
