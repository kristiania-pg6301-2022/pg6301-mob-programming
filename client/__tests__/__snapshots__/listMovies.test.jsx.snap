// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list movies should movie list 1`] = `"<div><h1>Movies from mFlix</h1><div><form><label>Country:<input id=\\"country-query\\" value=\\"\\"><button>Filter</button></label></form></div><h3>My movie</h3><img src=\\"https://example.com/poster1.jpg\\" width=\\"100\\" alt=\\"Movie poster\\"><div>Something happens</div><h3>My other movie</h3><img src=\\"https://example.com/void1.jpg\\" width=\\"100\\" alt=\\"Movie poster\\"><div>Nothing happens</div></div>"`;

exports[`list movies should show error message 1`] = `"<div class=\\"error-message\\"><h1>Error</h1><div>Error: Failed to fetch</div></div>"`;

exports[`list movies should show loading screen 1`] = `"<div class=\\"loading-indicator\\">Loading...</div>"`;
